Fix 764

mpoindexter opened the pull request:
When a zero time value is loaded from JSON or a DB in a way that assigns it the local timezone, it does not roudtrip in JSON as a value for which IsZero returns true. This causes KeyExpiry to be treated as a far past value instead of a nilish value.

See https://github.com/golang/go/issues/57040

<!--
Headscale is "Open Source, acknowledged contribution", this means that any
contribution will have to be discussed with the Maintainers before being submitted.

This model has been chosen to reduce the risk of burnout by limiting the
maintenance overhead of reviewing and validating third-party code.

Headscale is open to code contributions for bug fixes without discussion.

If you find mistakes in the documentation, please submit a fix to the documentation.
-->

<!-- Please tick if the following things apply. You… -->

- [x] have read the [CONTRIBUTING.md](./CONTRIBUTING.md) file
- [ ] raised a GitHub issue or discussed it on the projects chat beforehand
- [x] added unit tests
- [ ] added integration tests
- [ ] updated documentation if needed
- [ ] updated CHANGELOG.md

<!-- If applicable, please reference the issue using `Fixes #XXX` and add tests to cover your new code. -->
Fixes #764 

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
	- Introduced new test cases to enhance testing for time-sensitive operations and PostgreSQL interactions.
	- Added a function to allow users to specify timezones for container configurations.
	- Enhanced test configurability with new options for flexible testing scenarios.
  
- **Bug Fixes**
	- Resolved issues related to node registration in non-UTC timezones for PostgreSQL databases.

- **Documentation**
	- Updated CHANGELOG to reflect recent updates and fixes.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

kradalby commented:
Thank you, much appreciated, for this change I would need a integration test that first verify the problem, so something that when ran in sqlite/postgres mode returns pass/fail respectively, and then when this change is applied change to pass/pass.

If you have time to look at it would be great, or else I can see what I have time to do next week.

I recon that a fix for this issue also would merit a changelog entry.

mpoindexter commented:
The unit test I added would fail prior to this change, and now passes. It really actually has nothing to do with the database, but rather the `KeyExpiry` time object having a non-UTC timezone. Just the postgresql driver being used (pgx) happens to put all times into the local timezone, and the sqllite driver in use does not - but I guess that could change on the part of either driver since the the time represents the same instant of the zero value, so I think the test I added is the appropriate verification of this. If you disagree I can look at adding an integration test. I'll update the PR to add a changelog entry.

kradalby commented:
Integration tests to ensure full roundtrip via both drivers would be necessary. Both for this change and to ensure regressions, we have had libraries introduce silent failures before.

mpoindexter commented:
I guess it really depends on the timezone the tests run in - the existing integration tests should trip it as long as the timezone the headscale server runs in is non-UTC. I suspect on the integration tests the timezone is set to UTC and hence doesn't happen. I can take a look next week and see how hard it would be to introduce that, I'm not familiar with your integration test setup.

mpoindexter commented:
OK, added an integration test for this, as I suspected it was enough to ensure that the headscale container was in an alternate timezone using existing test scenarios. I added a test method that forces the container into an alternate TZ and runs it that fails without this change.  Also found that not all integration tests respected the env variable to use postgresql instead of sqlite, so fixed that as well, and added changelog entries.
